src/actions/pkg.cr

```diff
@@ -1,45 +1,129 @@
+        version = context.params.query["version"]?.try { |v| URI.unescape v }
+        # Iterate over the packages to find the relevant one.
+        selected_pkg = get_prefix_or_default(from: context).new_pkg package_name, version
+        raise NoSuchPackage.new context, package_name unless selected_pkg.exists?
+        # Stop here ^^ unless the package whose name was the :id URL parameter
+        # was found and we can query it
+        if key = context.params.query["get"]?
+          data = begin
+            selected_pkg.get_config key
+          rescue error : ConfigKeyError
+            raise NotFound.new context, cause: error
+          end
+          # A key was specified by the &get query parameter
+          build_json context.response do |json|
+            json.field package_name do
+              json.object do
+                json.field key, value: data
+              end
+            end
+          end
+          next context
+        else
+          # No specific key was requested, respond with all of the config
+          # options for this package. Additionally, respond with all
+          # dependent packages if the "libraries" boolean query parameter
+          # is specified.
+          build_json context.response do |builder|
+            build_config_response selected_pkg, builder
+            if parse_boolean_param("libraries", from: context) &&
+               (libs = selected_pkg.libs)
+              builder.field "libraries" do
+                libs.each { |pkg| build_config_response pkg, builder }
+              end
+            end
+          end
+        end
+        next context
+      end
+      raise Unauthorized.new context
+    end
+    # Delete a given package
+    relative_delete "/:id/delete" do |context|
+      if Actions.has_access? context, Access::Delete
+        package_name = URI.unescape context.params.url["id"]
+        selected_pkg = get_prefix_or_default(from: context).new_pkg package_name,
+          context.params.query["version"]?
+        raise NoSuchPackage.new context, package_name if selected_pkg.nil?
+        selected_pkg.delete confirmation: false { }
+        next context
+      end
+      raise Unauthorized.new context
+    end
+
+    # Build a package, returning the ID of the built image, and perhaps a status
+    # message? We could also use server-side events or a websocket to provide the
+    # status of this action as it occurs over the API, rather than just returning
+    # a result on completion.
+    #
+    # This route takes the optional query parameters "version" and "tag".
+    relative_post "/:id/build" do |context|
+      if Actions.has_access? context, Access::Create
+        package_name = URI.unescape context.params.url["package"]
+        pkg = get_prefix_or_default(from: context).new_pkg package_name,
+          version: context.params.query["version"]?
+        pkg.build confirmation: false { }
+        pkg.exists!
```

Comment:
I raises if the package exists, the method name is confusing. I should be `available!`
Anyway, that's the library job to do this checks.

Reply:
Currently, a package that already exists won't return an error, but it will for an application.

Reply:
ok. Is there any indication of failure for that method? Oh... I should be raising an `HTTPStatusError` from that block, shouldn't I?

Reply:
What block? This method (now named `available!`) will raise an exception

Reply:
I meant from `#build`, sorry

Reply:
Yes of course, there are lots of scenario for a build to fail.
The blocks is yields just before any change is done on the file-system. Whit the use of a boolean, we can choose to return a `400`, then after the block a `500`. What do you think?

Reply:
> The blocks is yield just before any change is done on the file-system

Please document these inconsistent uses of block yielding. I would say that even yielding to the block on an error is unidiomatic and should be documented.

> With the use of a boolean, we can choose to return a 400, then after the block a 500.

I'm sorry I don't understand what you mean by this

Reply:
This is documented

Reply:
Oh, I see... :man_facepalming:

Reply:
See https://github.com/DFabric/dppm/blob/master/src/prefix/pkg.cr#L109
> Build the package. Yields a block before writing on disk. When confirmation is set, the block must be true to continue.

The first stage is retrieving version, ensuring the packge exists etc.
If it fails, that's a `400 Bad Request`
Then, the block yields, and the second stage is executed, that performs changes on disk.
In any case of error, this is a `500 Internal Server Error`, because the build was supposed to work.

Reply:
I see. :+1:

Reply:
Ok...maybe I don't see. So... the block is yielded to if the file already exists. I get that. Then, you simulate a build, but only [`if confirmation`](https://github.com/DFabric/dppm/blob/6016bf9da7688dd02ce03557371cbd4eaacccfd4/src/prefix/program_data.cr#L86) (which `confirmation` is false in this case). When `confirmation` is false, you still yield to the block, though, just without [logging a bunch of stuff](https://github.com/DFabric/dppm/blob/6016bf9da7688dd02ce03557371cbd4eaacccfd4/src/prefix/program_data.cr#L87-L103).

It looks like,  as a library user, I have to *count* the number of times the block is called, and determine which error to throw based on said count?

I realize it's getting quite late in your timezone so we can put this off until tomorrow if you prefer.

Reply:
The block is always yielded, and only once.
So we can have:
```cr
build_init = false
pkg.build confirmation: false { build_init = true }
```

Reply:
We should also support simulate, later. The DPPM library isn't ready for that yet.

Reply:
This feature doesn't seem to work according to my test? Please see [the latest CI results](https://cloud.drone.io/DFabric/dppm-rest-api/214/1/5) when you get a chance.